# Since same-page merging makes heavy use of the mapping mechanism of the fiasco
# kernel, we need to increase the available kernel memory.
diff --git a/l4re/src/fiasco/src/kern/arm/config-arm.cpp b/l4re/src/fiasco/src/kern/arm/config-arm.cpp
index bc324994..60c5089c 100644
--- a/l4re/src/fiasco/src/kern/arm/config-arm.cpp
+++ b/l4re/src/fiasco/src/kern/arm/config-arm.cpp
@@ -41,7 +41,7 @@ public:
 
   enum
   {
-    KMEM_SIZE = 16 << 20,
+    KMEM_SIZE = 32 << 20,
   };
 
   // the default uart to use for serial console
# Tweak uvmm to be more lazy.
# We need to distinguish between read and write pagefaults here, otherwise same-
# page merging cannot happen.
diff --git a/l4re/src/l4/pkg/uvmm/server/src/ds_mmio_mapper.h b/l4re/src/l4/pkg/uvmm/server/src/ds_mmio_mapper.h
index 3f4c0e85..a53a9aa5 100644
--- a/l4re/src/l4/pkg/uvmm/server/src/ds_mmio_mapper.h
+++ b/l4re/src/l4/pkg/uvmm/server/src/ds_mmio_mapper.h
@@ -121,7 +121,8 @@ private:
     long res;
     l4_addr_t ls = local_start();
     // Make sure that the page is currently mapped.
-    res = page_in(ls + offset, true);
+    // TODO: This is an Arm-specific hack. Make ISA-independent!
+    res = page_in(ls + offset, vcpu.pf_write());

     if (res >= 0)
       {
# Increase heap size for the l4re_kernel.
# We need to do this here so that mmap() has enough memory to provide to the
# SimpleL4ReAllocator.
diff --git a/l4re/src/l4/pkg/l4re-core/l4re_kernel/server/src/mem.cc b/l4re/src/l4/pkg/l4re-core/l4re_kernel/server/src/mem.cc
index 05ac567a..ef377123 100644
--- a/l4re/src/l4/pkg/l4re-core/l4re_kernel/server/src/mem.cc
+++ b/l4re/src/l4/pkg/l4re-core/l4re_kernel/server/src/mem.cc
@@ -21,7 +21,7 @@
 #include "globals.h"
 
 enum {
-  Heap_max = L4_PAGESIZE * 64,
+  Heap_max = L4_PAGESIZE * 256,
 };
 
 extern char __executable_start[];
